File: Java-Kotlin/src/main/java/jianzhi/jz33.java
package jianzhi;

public class jz33 {

    // see jianzhi offer book
    public int GetUglyNumber_Solution(int index) {
        if (index <= 0) {
            return 0;
        }

        int[] uglyNums = new int[index];
        uglyNums[0] = 1;
        int nextUglyNumIdx = 1;
        int p2 = 0;
        int p3 = 0;
        int p5 = 0;
        while (nextUglyNumIdx < index) {
            int nextUglyNum = Math.min(uglyNums[p2] * 2, Math.min(uglyNums[p3] * 3, uglyNums[p5] * 5));
            uglyNums[nextUglyNumIdx] = nextUglyNum;
            while (uglyNums[p2] * 2 <= nextUglyNum) {
                p2++;
            }
            while (uglyNums[p3] * 3 <= nextUglyNum) {
                p3++;
            }
            while (uglyNums[p5] * 5 <= nextUglyNum) {
                p5++;
            }
            nextUglyNumIdx++;
        }

        return uglyNums[index - 1];
    }
}

File: Go-Solutions/nowcoder/jz33.go
package nowcoder

func GetUglyNumber_Solution(index int) int {
	if index <= 0 {
		return 0
	}

	uglyNums := make([]int, index)
	nextIdx := 1
	uglyNums[0] = 1
	p2, p3, p5 := 0, 0, 0
	for nextIdx < index {
		nextUglyNum := min33(uglyNums[p2]*2, min33(uglyNums[p3]*3, uglyNums[p5]*5))
		uglyNums[nextIdx] = nextUglyNum
		for uglyNums[p2]*2 <= nextUglyNum {
			p2++
		}
		for uglyNums[p3]*3 <= nextUglyNum {
			p3++
		}
		for uglyNums[p5]*5 <= nextUglyNum {
			p5++
		}
		nextIdx++
	}
	return uglyNums[index-1]
}

func min33(a, b int) int {
	if a < b {
		return a
	}
	return b
}
